### ISSUE TITLE
Tendon-Driven Actuation

### ISSUE THREAD

**Issue by TheFloatingString** [GazzolaLab/PyElastica#525]
**Is your feature request related to a problem? Please describe.**
I wanted to ask what would be the best way to implement tendon-driven actuation in PyElastica. There is growing interest in tendon-driven actuation for continuum robots (using "strings" to pull a continuum robot arc towards a desired pose). However, I can't find any existing implementation of tendons in PyElastica, and wanted to ask if it's best to (i) simulate tendons using a `RodBase` subclass, or to (ii) introduce a new class of Tendon-like objects that can be used for tendon actuation.

My intended use case is to use PyElastica as a high-fidelity simulation to validate tendon-driven models, and I think that evaluating tendon-driven components on PyElastica could benefit developers and researchers with similar interests. 

**Describe the solution you'd like**
- To determine what is the best way to implement tendon-driven actuation in PyElastica: (i) creating a new `RodBase` subclass or (ii) introducing a new `Tendon`-like class

**Describe alternatives you've considered**
- I considered creating multiple, thin rods (inherited from `elastica.rod.rod_base.RodBase`) that act as strings within a larger rod. 
- I also considered creating "virtual" tendons that are not actually rendered, but whose reaction forces are then applied as `EndpointForces` or `UniformTorques`. The limitation of this approach is that there is no actual rod, so the simulation would lose the ability to model tendon material properties and contact forces (i.e. friction) along the major rod

**Additional context**
- MuJoCo offers tendon-driven actuation (https://mujoco.readthedocs.io/en/stable/XMLreference.html#tendon)

**Comment by skim0119** [comment]
Hi -- Thank you for your interest in PyElastica.

With the limited information, I do not have a full picture of what type of robot and tendon model you are addressing.

For a general tendon-driven continuum robot, only the continuum robot is modeled with a Cosserat rod, and tendon actuation is expressed as an external force. Simulating stiff material, like non-biological engineered tendon (100 MPa ~ 10GPa), with PyElastica requires a smaller timestepper, and their shear/extension can easily be neglected, making full Cosserat rod simulation inefficient.

**Comment by TheFloatingString** [comment]
Hi @skim0119 , thank you so much for your help!

The type of tendon-driven model that I'm interested in is similar to a design proposed by Priyanka Rao et al. (2021), where tendons are inserted through small "holes" within a larger Cosserat-rode type structure. (It's hard to concisely describe in words, though Fig. 3 in their [2021 paper](https://www.frontiersin.org/journals/robotics-and-ai/articles/10.3389/frobt.2020.630245/full) represents the idea)

The design that I'm working with uses TPU, which could still fall within the range of Cosserat rod models. However, because the tendons are inserted inside holes inside the Cosserat rod, friction becomes a factor which is tough to accurately model using only external forces. 

Since my use case is to use PyElastica as a high-fidelity simulator to model the internal friction/deformation dynamics of such a tendon-driven rod, what would be the best practice with Elastica? (If I should create a new subclass of tendons that are inserted inside a Cosserat rod, or if there's a simpler approach)

Thanks so much for your help!

**Comment by skim0119** [comment]
There are many ways to go, and I'll respond based on my understanding.

> In short, contact + friction is never an easy problem for us as well. PyElastica is designed to provide flexibility such that any level of modeling could be done, but it becomes critical to understand the assumptions of the problem and limitations of the tool, especially for targeting fidelity.

The `high fidelity` in this problem seems not about the fidelity of the Cosserat rod, but the actuation model. It is two separate problems to solve at each time step: what load tendons are exerting, and the corresponding robot deformation. The latter part is what PyElastica `CosseratRod` would provide. The first part is purely up to your modeling choice.

### To solve the first part::

It appears that (correct me if I'm wrong), you want to add other rod models to represent tendon, and your exerting force (load+friction) is computed based on rod kinematic/dynamics interacting with the main (backbone) rod. What I would suggest, or maybe have it as a preliminary study, is to see if exerting force can be described in terms of external force that depends on position and rate (velocity). Friction can be modeled with some relaxation factor (force increase/decrease with incremental limit that depends on velocity) or by adding an arbitrary hysteresis effect. At some level, I would argue, these two approaches coincide.

Considering the "contact" -> "normal force" -> "friction estimation" route, you need to have your own model/equation for how the tendon's position and its motion exert the force and friction on the main body. Whether you use soft-contact with the Hertzian model, a hard boundary with zero penetration, or some linear spring added with friction, you will end up with a force equation and a mechanism of rate-based dissipation due to friction. Based on your choice of model and design of your robot, you can implement these forces in both the tendon and the main body. We provide a template and can provide help to integrate, but we decided not to maintain a list of implementations, since there are too many possible bespoke problem variations.

In my experience, there is no good or easy way of modeling once friction is in play, and high fidelity is only guaranteed once you have enough data to calibrate all the parameters. Hence, I alway recomend to start simple with fewer parameters to tune, and gradually adding complexity while checking the discrepancy. Most of the time, we found that more "complexity" in the model quickly makes the model itself useless (expansive, no intuition to guide, case-specific, etc.) compared to a simpler model.

### Some other thoughts

Considering the computational cost, fidelity, and validation difficulties, I would think for options to simplify the problem before moving towards full physical modeling.

- Is tendon position/orientation critical in your problem? If not, no need to concurrently simulate another rod equation. If yes, by how much?
- Do you have access to any experimental data? If not, what is your plan to justify the fidelity of the model? If yes, can you draw some data-driven intuition on any assumptions: inextensibility, characteristics on tendon-slippage, hysteresis, etc.
- Is speed your concern, or just the final configuration? --> quasi-static?
- How many hyperparameters would play in the model

### Some other comments/concerns

- Typical tendon I imagine, like some sort of wires in plastic or metal or composite, are still very stiff materials that makes rod equation very slow to resolve.
- Cosserat rod model in PyElastica is "mathematically" sensible for the incompressible (poisson=0.5) case; in other words, the change in radius for compressible material is likely not accurate. Typically it is negligible for soft material, where deformation is significant, but it becomes sensitive for contact modeling.

**Comment by TheFloatingString** [comment]
Hi @skim0119 , thank you so much for the kind reply! I'm so sorry for the delay, this is super helpful.

Indeed, it sounds like my case requires decoupling the simulation into two problems: the actuation model and the corresponding deformation simulation. My initial blocker was related to how to properly measure the effect of friction between the tendon and the parent rod's body. After some thought, it makes sense that the rod actuation can be represented by forces, and the deformation aspect should then be simulated using PyElastica's rod assets. 

I was initially also curious about how accurately a simulator's friction model mirrors the real world friction response. Though it sounds like real-world friction is quite non-linear, and collecting real-world data to then fine-tune the simulator's friction parameters makes sense.

Thank you so much for your time and help, it's super appreciated!

**Comment by skim0119** [comment]
@TheFloatingString I'll leave some of my thoughts.

## digression on friction
In my opinion any "dissipation" related modeling should start with data or experiment first. Systematically compare with progressively non-linear model: no-friction, simple-linear, and rate-dependent, etc. As analyzing the differences, one can setup an appropriate model. It is hard to discuss "accuracy" of simulated friction against real-world friction, because there are plethora of cause behind the friction between surfaces in real-world. In some sense, numerical simulation is best accurate up to what it models; model has its own discrepancy to the real world.

This view-point brings in a circular argument: it is accurate because one can write the model/equation non-linear enough to fit the data. Hence, it is up for user to balance the _complexity_ of model/equation one wants to write versus the _explainability_ or _interpretability_ of the model.

`PyElastica` is designed such that the framework is capable to include various type of friction or dissipating model, but exactly what model/equation to use is up to user to decide and implement. We have simple linear friction examples within the `contact` modules, hopefully gives a template for people to understand the structure and adopt to other type of frictions.

**Comment by TheFloatingString** [comment]
@skim0119 thank you so much for your help! 

Agreed that real-world friction is challenging to properly characterize. I was initially blocked on how best to characterize the friction response, though it makes sense now that it's best to first determine the real-world friction and adapt a friction model subsequently.

Really appreciate all your time and help!
